x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -1365,23 +1385,74 @@ private void unlink(final LFUCacheEntry entry) {
 
         /**
          * Cycles through the {@link LFUCacheEntry} from 0 to max frequency and
-         * tries to evict a chunk if no one is holding onto its resources anymore
+         * tries to evict a chunk if no one is holding onto its resources anymore.
+         *
+         * Also regularly polls for free regions and thus might steal one in case any become available.
          *
-         * @return the frequency of the evicted entry as integer or -1 if no entry was evicted from cache
+         * @return a now free IO region or null if none available.
          */
-        private int maybeEvict() {
+        private SharedBytes.IO maybeEvictAndTake(Runnable evictedNotification) {
             assert Thread.holdsLock(SharedBlobCacheService.this);
-            for (int currentFreq = 0; currentFreq < maxFreq; currentFreq++) {
-                for (LFUCacheEntry entry = freqs[currentFreq]; entry != null; entry = entry.next) {
-                    boolean evicted = entry.chunk.tryEvict();
-                    if (evicted && entry.chunk.io != null) {
-                        unlink(entry);
-                        keyMapping.remove(entry.chunk.regionKey, entry);
-                        return currentFreq;
+            long currentEpoch = epoch.get();
+            SharedBytes.IO freq0 = maybeEvictAndTakeForFrequency(evictedNotification, 0);
```

Comment:
I wonder if it still makes sense to add new entries with `freq=1`? Once the cache will be populated with `epoch=0,freq=1` entries the next cache miss will schedule a decay computation that will  bring back all (or all but one) entries back to freq=0?

Reply:
I pondered on that too and I am open to putting it to level 2 instead. I left it at 1 since i think it might not be too important though, given that a new entry that is unfortunate to decay to level 0 quickly will be towards the end of the freq=1 LRU chain and thus should have a good chance of being touched again in the new epoch. I'll add a todo for now.

Reply:
Thanks for adding the TODO. I was also worried on the time taken for the first decay computation but the additionnal logging in 104784 should tell me how long it takes.

That being said freq=2 makes sens too. I also think we could test different initial frequencies according to other information, like the file read or the blob region (ie, freq=2 for the first blob region, or blob region containing cfs, and leave 1 for the rest etc)